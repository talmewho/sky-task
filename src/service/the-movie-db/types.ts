type SummarizedContent = {
  id: number,
  name: string,
  type: 'tv' | 'movie',
  year: number
};

export type Cast = {
  id: number,
  name: string,
  imageURL: string,
  biography: string,
  contents: SummarizedContent[],
  age?: number
};

export type Content = SummarizedContent & {
  imageURL: string,
  overview: string,
  cast: {id: number, name: string}[],
  year: number
};

export type SearchResult = {
  id: number,
  name: string,
  imageURL: string,
  knownFor?: string[],
  type: 'tv' | 'movie' | 'person'
};

export type SearchResults = {
  results: SearchResult[],
  totalCount: number,
  totalPageCount: number,
  page: number
};
