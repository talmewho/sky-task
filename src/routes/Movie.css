.movie-details {
  display: flex;
}

.movie-details .hero {
  margin: 0 1rem;
}

.movie-details .content-title {
  margin-top: 0;
}

.movie-details .content-description {
  margin: 1rem 0;
}

.movie-details .hero img {
  width: 375px;
}