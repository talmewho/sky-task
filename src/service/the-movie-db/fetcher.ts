import {Content, Person, SearchResult, SearchResults, Suggestion} from './types';
import {
  TheMovieDBPersonShowCredit,
  TheMovieDBPersonMovieCredit,
  TheMovieDBShowCastCredit,
  TheMovieDBMovieCastCredit,
  TheMovieDBPersonAndCredits,
  TheMovieDBShowSearchResult,
  TheMovieDBMovieSearchResult,
  TheMovieDBPersonSearchResult,
  TheMovieDBSearchResult
} from '../../data-access/TheMovieDB.types';

import constants from '../../common/TheMovieDB.constants';

import configuration from '../../configuration';

import TheMovieDBFetcher from '../../data-access/TheMovieDBFetcher';


const defaultFetcher = new TheMovieDBFetcher();

type GetShowOptions = {
  showID: string,
  fetcher?: TheMovieDBFetcher,
  posterImageSize: string
};

type GetMovieOptions = {
  movieID: string,
  fetcher?: TheMovieDBFetcher,
  posterImageSize: string
};

type getPersonOptions = {
  personID: string,
  fetcher?: TheMovieDBFetcher,
  profileImageSize: string
};

type SearchOptions = {
  query: string,
  page: number,
  fetcher?: TheMovieDBFetcher,
  profileImageSize: string,
  posterImageSize: string
};

type GetSuggestionsOptions = {
  query: string,
  fetcher?: TheMovieDBFetcher
};


const getImageURL =
  async (
    relativeURL: string | null,
    type: 'poster_sizes' | 'profile_sizes',
    size: string,
    fallbackURL: string,
    fetcher: TheMovieDBFetcher) => {
  try {
    return relativeURL ?
             await fetcher.getImageURL(relativeURL, type, size) :
             fallbackURL;
  } catch (e) {
    return fallbackURL;
  }
};

const getYear = (date: string): number => {
  return date ? new Date(date).getFullYear() : 0
};

export const getShow =
  async ({showID, posterImageSize, fetcher = defaultFetcher}: GetShowOptions): Promise<Content> => {
  const {id, name, poster_path, first_air_date: releaseDate, overview, credits: {cast}} =
    await fetcher.getShow(showID);

  const year = getYear(releaseDate);

  const imageURL =
    await getImageURL(
      poster_path,
      constants.sizeField.poster,
      posterImageSize,
      configuration.defaultContentImageURL,
      fetcher
    );


  return {
    id,
    name,
    year,
    type: constants.mediaType.tv,
    overview,
    imageURL,
    cast: cast.map(({name, id}: TheMovieDBShowCastCredit) => ({name, id}))};
};

export const getMovie =
  async ({movieID, posterImageSize, fetcher = defaultFetcher}: GetMovieOptions): Promise<Content> => {
  const {id, title: name, poster_path, release_date, overview, credits: {cast}} =
    await fetcher.getMovie(movieID);

  const year = getYear(release_date);

  const imageURL =
    await getImageURL(
      poster_path,
      constants.sizeField.poster,
      posterImageSize,
      configuration.defaultContentImageURL,
      fetcher
    );

  return {
    id,
    name,
    year,
    type: constants.mediaType.movie,
    overview,
    imageURL,
    cast: cast.map(({name, id}: TheMovieDBMovieCastCredit) => ({name, id}))};
};

export const getPerson =
  async ({personID, profileImageSize, fetcher = defaultFetcher}: getPersonOptions): Promise<Person> => {
  const {id, name, profile_path, birthday, deathday, biography, combined_credits: {cast}}: TheMovieDBPersonAndCredits =
    await fetcher.getPerson(personID);

  const age =
    new Date(
      (new Date(deathday || Date.now()).getTime()) -
      (new Date(birthday)).getTime()
    ).getUTCFullYear() - 1970;

  const imageURL =
    await getImageURL(
      profile_path,
      constants.sizeField.profile,
      profileImageSize,
      configuration.defaultPersonImageURL,
      fetcher
    );

  const contents = cast.map((credit: TheMovieDBPersonShowCredit | TheMovieDBPersonMovieCredit) => {
    let name: string;
    let year: number;
    if (credit.media_type === constants.mediaType.tv) {
      const showCredit = credit as TheMovieDBPersonShowCredit;
      name = showCredit.name;
      year = getYear(showCredit.first_air_date);
    } else {
      const movieCredit = credit as TheMovieDBPersonMovieCredit;
      name = movieCredit.title;
      year = getYear(movieCredit.release_date);
    }
    return {id: credit.id, name, type: credit.media_type, year};
  })
    .sort(({year}, {year: year2}) => (year2 - year))
    // Removing duplicates, assumes a sorted array.
    .filter(({id}, index, array) => {
      if (array[index - 1] && array[index - 1].id === id) {
        return false;
      }
      return true;
    });

  return {
    imageURL,
    id,
    name,
    age,
    biography,
    contents
  };
};

const getKnownFor = (
  rawResults: TheMovieDBSearchResult[],
  posterImageSize: string,
  profileImageSize: string,
  fetcher: TheMovieDBFetcher
): Promise<SearchResult>[] =>
  rawResults.map<Promise<SearchResult>>(async (item: TheMovieDBSearchResult): Promise<SearchResult> => {
    let name;
    let year;
    let knownFor: string[] | undefined = undefined;
    let imageURL: string;
    if (item.media_type === constants.mediaType.tv) {
      const show = item as TheMovieDBShowSearchResult;
      name = show.name;
      year = getYear(show.first_air_date);
      imageURL =
        await getImageURL(
          show.poster_path,
          constants.sizeField.poster,
          posterImageSize,
          configuration.defaultContentImageURL,
          fetcher
        );
    } else if (item.media_type === constants.mediaType.movie) {
      const movie = item as TheMovieDBMovieSearchResult;
      name = movie.title;
      year = getYear(movie.release_date);
      imageURL =
        await getImageURL(
          movie.poster_path,
          constants.sizeField.poster,
          posterImageSize,
          configuration.defaultContentImageURL,
          fetcher
        );
    } else {
      const person = item as TheMovieDBPersonSearchResult;
      name = person.name;
      imageURL =
        await getImageURL(
          person.profile_path,
          constants.sizeField.profile,
          profileImageSize,
          configuration.defaultPersonImageURL,
          fetcher
        );

      knownFor = person.known_for.slice(0, 3).map(content => {
        return (content as TheMovieDBShowSearchResult).name ||
               (content as TheMovieDBMovieSearchResult).title;
      });
    }
    return {name, id: item.id, year, type: item.media_type, knownFor, imageURL};
  });

export const search =
  async ({
    query,
    page,
    posterImageSize,
    profileImageSize,
    fetcher = defaultFetcher
  }: SearchOptions): Promise<SearchResults> => {
  const {
    results: rawResults,
    total_results: totalCount,
    total_pages: totalPageCount,
    page: outputPage
  } = await fetcher.search(query, page);

  const results: SearchResult[] =
    await Promise.all(getKnownFor(rawResults, posterImageSize, profileImageSize, fetcher));
  return {results, totalCount, page: outputPage, totalPageCount};
};

export const getSuggestions =
  async ({query, fetcher = defaultFetcher}: GetSuggestionsOptions): Promise<Suggestion[]> => {
  const {results} = await fetcher.search(query, 1);
  return results.map(item => {
    const name =
      ((item as (TheMovieDBShowSearchResult | TheMovieDBPersonSearchResult)).name) ||
      ((item as TheMovieDBMovieSearchResult).title);
    return {name, id: item.id, type: item.media_type};
  });
};
