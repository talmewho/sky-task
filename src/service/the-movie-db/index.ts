import * as fetcher from './fetcher';
export {fetcher}
