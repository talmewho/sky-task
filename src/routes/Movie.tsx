import React, {useState, useEffect} from 'react';
import {useParams, Link} from 'react-router-dom';

import {Content} from '../service/the-movie-db/types';

import {fetcher} from '../service/the-movie-db';

import './Movie.css';

const Movie: React.FC = () => {
  const {id} = useParams();
  const [movie, setMovie] = useState<Content>();

  useEffect(() => {
    (async () => {
      setMovie(await fetcher.getMovie({movieID: id, posterImageSize: 'w342'}));
    })();
  }, [id]);

  if (!movie) {
    return (<div>Loading...</div>);
  }

  let cast: React.ReactNode;
  if (movie.cast.length) {
    const names = movie.cast.map(({id, name}, index, array) => {
      const comma: string =
        index === array.length - 2 ?
          ' and ' :
        index < array.length - 2 ?
          ', ' :
          '.';
      return (
        <React.Fragment key={id}>
          <Link to={`/cast/${id}`}>{name}</Link>{comma}
        </React.Fragment>
      );
    });
    cast = (
      <>
        {'Starring '}
        {names}
      </>
    );
  }

  const year = movie.year ? ` (${movie.year})` : '';

  return (
    <article className="movie-details">
      <figure className="hero"><img src={movie.imageURL} alt={movie.name} /></figure>
      <section className="details">
        <h1 className="content-title">{movie.name}{year}</h1>
        <div className="content-description">{movie.overview}</div>
        {cast}
      </section>
    </article>
  );
};

export default Movie;
