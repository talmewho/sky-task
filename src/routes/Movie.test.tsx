import React from 'react';
import renderer, {act} from 'react-test-renderer';

import getMovie1Mock from './mocks/movie-1';

import {getMovie} from '../service/the-movie-db/fetcher';

import Movie from './Movie';

jest.mock('react-router-dom', () => {
    return {useParams: () => ({id: '1'}), Link: ({children, to}: {
        children: React.ReactChildren,
        to: string
    }) => <a href={to}>{children}</a>};
});

jest.mock('../service/the-movie-db/fetcher');
const mockGetMovie = getMovie as jest.MockedFunction<typeof getMovie>;

const waitForUpdate = async () => {
    return act(async () => await new Promise((resolve) => setTimeout(resolve, 0)));
};

describe('Components.Movie', () => {
    beforeEach(() => {
        mockGetMovie.mockClear();
    });

    afterEach(async () => {
        await waitForUpdate();
    });

    it('shows a loading indicator before data is fetched', async () => {
        const component = renderer.create(<Movie />);
        await waitForUpdate();
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('data is fetched and shown', async () => {
        mockGetMovie.mockResolvedValueOnce(getMovie1Mock());

        const component = renderer.create(<Movie />);
        await waitForUpdate();

        expect(mockGetMovie).toBeCalledTimes(1);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('data fetching throws, error is shown', async () => {
        mockGetMovie.mockRejectedValue(new Error(''));

        const component = renderer.create(<Movie />);
        await waitForUpdate();

        expect(mockGetMovie).toBeCalledTimes(1);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
