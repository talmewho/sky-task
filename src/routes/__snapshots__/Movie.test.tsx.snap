// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components.Movie data fetching throws, error is shown 1`] = `
<div>
  Could not load the movie. :( Try again later.
</div>
`;

exports[`Components.Movie data is fetched and shown 1`] = `
<article
  className="movie-details"
>
  <figure
    className="hero"
  >
    <img
      alt="Amélie"
      src="https://image.tmdb.org/t/p/w342/slVnvaH6fpF22154vnjQJdXCVZd.jpg"
    />
  </figure>
  <section
    className="details"
  >
    <h1
      className="content-title"
    >
      Amélie
       (2001)
    </h1>
    <div
      className="content-description"
    >
      At a tiny Parisian café, the adorable yet painfully shy Amélie accidentally discovers a gift for helping others. Soon Amelie is spending her days as a matchmaker, guardian angel, and all-around do-gooder. But when she bumps into a handsome stranger, will she find the courage to become the star of her very own love story?
    </div>
    Starring 
    <a
      href="/person/2405"
    >
      Audrey Tautou
    </a>
     and 
    <a
      href="/person/2406"
    >
      Mathieu Kassovitz
    </a>
    .
  </section>
</article>
`;

exports[`Components.Movie shows a loading indicator before data is fetched 1`] = `
<div>
  Loading...
</div>
`;
